pm: Dynamically bind a device to a power domain

ceolin opened the pull request:
Add API to add devices to a power domain in runtime. The number of
devices that can be added is defined in build time.

The script gen_handles.py will check the number defined in
`CONFIG_PM_DEVICE_POWER_DOMAIN_DYNAMIC` to resize the handles vector,
adding empty slots in the supported sector to be used later.

v2:

- Only resize power domains (instead of all devices when the option is enabled)
- ~~The changes in the linker script is in one separated file now~~
- Add a hidden option to simplify the `ifdef` logic
- Add an api to remove a device from a power domain (the counter part of the `add` function)

Commit:
tests: pm: Add a power domain test

Add simple test to exercise power domain behavior
with device runtime.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: device: Dynamically add a device to a power domain

Add API to add devices to a power domain in runtime. The number of
devices that can be added is defined in build time.

The script gen_handles.py will check the number defined in
`CONFIG_PM_DEVICE_POWER_DOMAIN_DYNAMIC` to resize the handles vector,
adding empty slots in the supported sector to be used later.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

gmarull reviewed (commented):
I'd really like to understand the use case for this, since all domain information should be available at build time. Zephyr doesn't support _dynamic devices_, so why should a feature like this be needed?

tejlmand commented on scripts/gen_handles.py:
```diff
@@ -62,6 +62,8 @@ def parse_args():
 
     parser.add_argument("-k", "--kernel", required=True,
                         help="Input zephyr ELF binary")
+    parser.add_argument("-d", "--dynamic-devices", required=True,
+                        help="Input number of dynamic devices allowed")
```
why does this need to be a required argument ?

Why not make it optional, and if not provided treat it as `0`.

ceolin commented:
> I'd really like to understand the use case for this, since all domain information should be available at build time. Zephyr doesn't support _dynamic devices_, so why should a feature like this be needed?

That is true, we can't instantiate device at runtime, but tthis is usefeul when one wants to re-use the same firmware in slightly different targets and just  bind a device to a power domain  during the system intialization.

JordanYates commented on subsys/pm/device.c:
```diff
@@ -101,6 +101,57 @@ int pm_device_action_run(const struct device *dev,
+			       const struct device *domain)
+{
+#ifdef CONFIG_PM_DEVICE_POWER_DOMAIN
+	device_handle_t *rv = (device_handle_t *)domain->handles;
```
Isn't this line discarding the `const` qualifier from the handles?
https://github.com/zephyrproject-rtos/zephyr/blob/a5b33f89b729d5803e3a6879d89eaef3b98a5ebd/include/device.h#L468
AFAIK the handles currently live in ROM, so I'm not sure how this can work

ceolin replied:
It is resized in compile time, so the size is pre-allocated.

ceolin replied:
You're completely right, I've tested in the integration test (that is using native posix) and didn't catch it.

ceolin replied:
So, I just did some small changes, when this option is enabled, the array is put in the ram instead of the rom. It is indeed memory cost, but it is a price to pay for it. If this option  is not set, anything changes and should not be any impact. A future change may put just an additional array in RAM and keep the ones know at build time in ROM. This change though, will not require any API change and will be done entirely internally.

JordanYates reviewed (changes requested):
I don't believe the current approach can work without moving all handle arrays from ROM to RAM, which will be a significant memory cost.

JordanYates commented on subsys/pm/device.c:
```diff
@@ -101,6 +101,57 @@ int pm_device_action_run(const struct device *dev,
+			       const struct device *domain)
+{
+#ifdef CONFIG_PM_DEVICE_POWER_DOMAIN
+	device_handle_t *rv = (device_handle_t *)domain->handles;
```
The explicit `(device_handle_t *)` is what is making you miss the qualifier problems, the compiler isn't complaining when you drop the `const` qualifier. When this is correct you won't need the explicit cast.

ceolin commented on include/device.h:
```diff
@@ -465,7 +465,7 @@ struct device {
 	 * device_required_handles_get().
 	 */
-	const device_handle_t *const handles;
+	const device_handle_t *handles;
```
argh, I dropped the wrong the `const`. Somehow in mind we were change the pointer to point a const device. But that is right, the pointer is constant.

JordanYates commented:
> > I'd really like to understand the use case for this, since all domain information should be available at build time. Zephyr doesn't support _dynamic devices_, so why should a feature like this be needed?
> 
> That is true, we can't instantiate device at runtime, but tthis is usefeul when one wants to re-use the same firmware in slightly different targets and just bind a device to a power domain during the system intialization.

Isn't that the purpose of board revisions?

ceolin commented:
> > > I'd really like to understand the use case for this, since all domain information should be available at build time. Zephyr doesn't support _dynamic devices_, so why should a feature like this be needed?
> > 
> > 
> > That is true, we can't instantiate device at runtime, but tthis is usefeul when one wants to re-use the same firmware in slightly different targets and just bind a device to a power domain during the system intialization.
> 
> Isn't that the purpose of board revisions?

No, the use case here is to use the same binary in different targets.

JordanYates commented:
> No, the use case here is to use the same binary in different targets.

Right, but my point is that if they are different targets, why are they not board revisions with the appropriate modifications in overlay files? Why are we doing this hackery, when the problem is already solvable with Zephyr's existing mechanism for different targets.

ceolin commented:
> > > I'd really like to understand the use case for this, since all domain information should be available at build time. Zephyr doesn't support _dynamic devices_, so why should a feature like this be needed?
> > 
> > 
> > That is true, we can't instantiate device at runtime, but tthis is usefeul when one wants to re-use the same firmware in slightly different targets and just bind a device to a power domain during the system intialization.
> 
> Isn't that the purpose of board 

> > No, the use case here is to use the same binary in different targets.
> 
> Right, but my point is that if they are different targets, why are they not board revisions with the appropriate modifications in overlay files? Why are we doing this hackery, when the problem is already solvable with Zephyr's existing mechanism for different targets.

This require to build two images, the use case we have involves one single image that can be  deployed in slightly different targets. Of course, it is possible to have two images and distribute them to the right target, but that is not the use case we have.

JordanYates commented:
> This require to build two images, the use case we have involves one single image that can be deployed in slightly different targets. Of course, it is possible to have two images and distribute them to the right target, but that is not the use case we have.

I understand that.

Is it worthwhile and desired to support this feature, when no other part of Zephyr works this way.
Why is it possible to do this with power domains, but not SPI or I2C pins? Would support for those be merged if someone asked?

Unless the answer to the above is yes, my feeling is that it would be very inconsistent to have this one corner of Zephyr work in this way, but nothing else. But these are larger questions better answered by dev-review or the TSC IMO.

JordanYates commented:
Of course there is nothing stopping you from writing another version of `power_domain_gpio` which has a custom API for dynamically adding dependencies. I would have no problems with that.

ceolin commented:
> > This require to build two images, the use case we have involves one single image that can be deployed in slightly different targets. Of course, it is possible to have two images and distribute them to the right target, but that is not the use case we have.
> 
> I understand that.
> 
> Is it worthwhile and desired to support this feature, when no other part of Zephyr works this way? Why is it possible to do this with power domains, but not SPI or I2C pins? Would support for those be merged if someone asked?
> 
> Unless the answer to the above is yes, my feeling is that it would be very inconsistent to have this one corner of Zephyr work in this way, but nothing else. But these are larger questions better answered by dev-review or the TSC IMO.

If someone asked and had a use case and the implementation wouldn't  impact others use case, much likely yes ...
That is not a big feature nor something that has a toll in anything, if things like that needs to go to TSC or dev-meeting my understand is that everything has to pass trough meetings ....

JordanYates commented:
> That is not a big feature nor something that has a toll in anything, if things like that needs to go to TSC or dev-meeting my understand is that everything has to pass trough meetings ....

It's not a big feature, but it does go against the current model of "Hardware relationships are defined in devicetree".
Maybe that's not worthy of a trip to dev-review, maybe it is, it's not my call, I'm just a lowly collaborator :)

ceolin commented:
> > That is not a big feature nor something that has a toll in anything, if things like that needs to go to TSC or dev-meeting my understand is that everything has to pass trough meetings ....
> 
> It's not a big feature, but it does go against the current model of "Hardware relationships are defined in devicetree". Maybe that's not worthy of a trip to dev-review, maybe it is, it's not my call, I'm just a lowly collaborator :)

that is not the common for sure, but there is, at lest, one other case 

https://docs.zephyrproject.org/latest/guides/pinctrl/index.html?highlight=one%20effects#dynamic-pin-control

```
Dynamic pin control refers to the capability of changing pin configuration at runtime. This feature can be useful in situations where the same firmware needs to run onto slightly different boards, each having a peripheral routed at a different set of pins. 
```

gmarull reviewed (dismissed):
This PR introduces the concept of _dynamic devices_, something we don't have in Zephyr. With current Zephyr design, everything _gets in_ at compile time, so trying to do dynamic stuff has chances to require hacks. For example, unlike Linux, there's no Devicetree at runtime.

We have the concept of board revisions to handle hardware variations. If that option doesn't solve the problem you're attempting to solve with this PR, we should first understand why.

If we want to open the door to dynamic stuff, I think we should have a broader discussion. It has many implications, for example, whether the current Devicetree approach makes sense or not. 

The pinctrl dynamic option that was added because it can cover some edge cases, and it's super simple: it's just dropping `const` qualifier and a pointer swap, really. It doesn't for example require changing core scripts. Its usage, scope, and capabilities are clearly documented so that we avoid people abusing the functionality.

So to summarize, I think we should have a discussion in API or dev-review before introducing this sort of functionality.

nashif commented:
> The pinctrl dynamic option that was added because it can cover some edge cases, and it's super simple: it's just dropping `const` qualifier and a pointer swap, really. It doesn't for example require changing core scripts. Its usage, scope, and capabilities are clearly documented so that we avoid people abusing the functionality.

This is EXACTLY the same case. 

The end goal here was NOT to introduce dynamic devices, it is about assigning devices to a power domain at runtime, based on events or hardware discovery that can only happen at runtime and It happens that we use devices for managing power domain. 

We also want to for example manage device dependencies and extend what we get from device tree at runtime and feed this to power management for device power management. This is similar. DT device dependencies was always just the start and more resources will come in to be managed during the powerup and powerdown process.

> We have the concept of board revisions to handle hardware variations. If that option doesn't solve the problem you're attempting to solve with this PR, we should first understand why.

That was explained above already. There are cases where differences between HW variants do not justify a new firmware image and the process associated with managing multiple FW images.

gmarull commented:
> > The pinctrl dynamic option that was added because it can cover some edge cases, and it's super simple: it's just dropping `const` qualifier and a pointer swap, really. It doesn't for example require changing core scripts. Its usage, scope, and capabilities are clearly documented so that we avoid people abusing the functionality.
> 
> This is EXACTLY the same case.

No, it's not EXACTLY the same. This PR is hacking into core scripts, pinctrl doesn't go that far.

> 
> The end goal here was NOT to introduce dynamic devices, it is about assigning devices to a power domain at runtime, based on events or hardware discovery that can only happen at runtime and It happens that we use devices for managing power domain.

Could we have concrete examples, as we did for pinctrl? There's even a sample that was presented to showcase it. This is not happening here. I think it would help to make progress here.

ceolin commented:
> 



> > > The pinctrl dynamic option that was added because it can cover some edge cases, and it's super simple: it's just dropping `const` qualifier and a pointer swap, really. It doesn't for example require changing core scripts. Its usage, scope, and capabilities are clearly documented so that we avoid people abusing the functionality.
> > 
> > 
> > This is EXACTLY the same case.
> 
> No, it's not EXACTLY the same. This PR is hacking into core scripts, pinctrl doesn't go that far.
> 

That is the SAME use case, not the same implementation. And lets be honest:
```
One of the effects of enabling dynamic pin control is that pinctrl_dev_config will be stored in RAM
instead of ROM (not states or pin configurations, though). The user can then use pinctrl_update_states()
to update the states stored in pinctrl_dev_config with a new set. This effectively means that the device
driver will apply the pin configurations stored in the updated states when it applies a state.
```
So, it was not needed to go that far because the whole thing was stored in RAM to support the "SAME" use case. This implementation is this way to not have any impact for those who don't need it.

> > The end goal here was NOT to introduce dynamic devices, it is about assigning devices to a power domain at runtime, based on events or hardware discovery that can only happen at runtime and It happens that we use devices for managing power domain.

Exactly ... The only thing that is dynamic is binding existent devices to existent domains.

> 
> Could we have concrete examples, as we did for pinctrl? There's even a sample that was presented to showcase it. This is not happening here. I think it would help to make progress.

There is a test for this use case, it is not like we have real samples for all features that are provided. For example your pr https://github.com/zephyrproject-rtos/zephyr/pull/42040 introduced new API / feature and it was validated with tests and not samples in real product. That is understandable, there are people using it in products that can't be disclosed.

nashif commented:
> #42020 introduces a similar API to Linux PM QoS (simpler and part of policy manager because of Zephyr design), so it's not anything new in the end.

Confused how #42020 is relevant here.

> Regarding dynamic pinctrl, it was added because a few people asked for it multiple times, so there was an interest in the community.

aha, and what part of this PR is not "interest in the community"?


> Regarding the proposal in this PR, I think an example would help. Samples don't need to show any private or confidential information, just illustrate a case where it is clear that this feature is useful and that can't be achieved in any other way, not doing that because this is only in `products that can't be disclosed` doesn't sound really convincing.

You have two SKUs that share 99.9% of the HW but assume there are some differences in how certain devices are connected to power, so device A which is part of domain A in SKU 1 is part of domain B in SKU 2. You can discover what SKU you are running on and adjust things accordingly at runtime. This is not something anyone would want to spend time on writing a sample, such sample wont make any sense for Zephyr and I would personally reject it.

It should be enough to describe this in the documentation.

ceolin commented:
> > There is a test for this use case, it is not like we have real samples for all features that are provided. For example your pr #42040 introduced new API / feature and it was validated with tests and not samples in real product. That is understandable, there are people using it in products that can't be disclosed.
> 
> #42020 introduces a similar API to Linux PM QoS (simpler and part of policy manager because of Zephyr design), so it's not anything new in the end.

Really ? https://github.com/torvalds/linux/blob/master/include/linux/pm_domain.h#L243
This is supported by Linux API as well.

> 
> Regarding dynamic pinctrl, it was added because a few people asked for it multiple times, so there was an interest in the community. It's also there because it was easy to implement without too much hackery, and the scope of the change was in the subsystem itself (`pinctrl.h`). I personally had no much interest on it, because in the end Zephyr doesn't have a dynamic device model so one is fairly limited on things that can be done at runtime. Also, it doesn't solve all cases: it is only guaranteed to work on non-initialized devices (see we still have #11918 open).
> 
> Regarding the proposal in this PR, I think an example would help. Samples don't need to show any private or confidential information, just illustrate a case where it is clear that this feature is useful and that can't be achieved in any other way, not doing that because this is only in `products that can't be disclosed` doesn't sound really convincing. Tests just show that the API works, but other than that don't tell me how this is expected to be used or what is the usecase (for #42020 one could refer e.g. to [these slides](https://events.static.linuxfound.org/images/stories/pdf/lceu2012_wysocki.pdf) or [kernel docs](https://www.kernel.org/doc/html/latest/power/pm_qos_interface.html)).

The use case is pretty simple,  the information about which power domain a device belongs is not know at build time, and a lookup has to be done at runtime figure it out and bind the device to a domain. Write a simple sample that does it would be that different from the test ? Much likely not, this would make the review different, certainly not !
Regarding #42020, it could refer but it didn't. Also point to Linux is not right approach, and one could ask the use case on Zephyr (not telling that there is not use case, just to be clear). And if points to Linux is enough to justify things, you are going to find information about generic power domain and capability to add/remove devices at runtime in the slides you have linked.

ceolin commented:
> > #42020 introduces a similar API to Linux PM QoS (simpler and part of policy manager because of Zephyr design), so it's not anything new in the end.
> 
> Confused how #42020 is relevant here.

Just to point out that the criteria used to reject this PR was not applied in that one.

> 
> > Regarding dynamic pinctrl, it was added because a few people asked for it multiple times, so there was an interest in the community.
> 
> aha, and what part of this PR is not "interest in the community"?
> 
> > Regarding the proposal in this PR, I think an example would help. Samples don't need to show any private or confidential information, just illustrate a case where it is clear that this feature is useful and that can't be achieved in any other way, not doing that because this is only in `products that can't be disclosed` doesn't sound really convincing.
> 
> You have two SKUs that share 99.9% of the HW but assume there are some differences in how certain devices are connected to power, so device A which is part of domain A in SKU 1 is part of domain B in SKU 2. You can discover what SKU you are running on and adjust things accordingly at runtime. This is not something anyone would want to spend time on writing a sample, such sample wont make any sense for Zephyr and I would personally reject it.
> 
> It should be enough to describe this in the documentation.

nashif commented:
> > > #42020 introduces a similar API to Linux PM QoS (simpler and part of policy manager because of Zephyr design), so it's not anything new in the end.
> > 
> > 
> > Confused how #42020 is relevant here.
> 
> Just to point out that the criteria used to reject this PR was not applied in that one.

oh, you mean https://github.com/zephyrproject-rtos/zephyr/pull/42040, @gmarull  linked the wrong pr

gmarull commented:
> > #42020 introduces a similar API to Linux PM QoS (simpler and part of policy manager because of Zephyr design), so it's not anything new in the end.
> 
> Confused how #42020 is relevant here.

It introduces a new API, with no in-tree users (yet), that's the only relation really.

> 
> > Regarding dynamic pinctrl, it was added because a few people asked for it multiple times, so there was an interest in the community.
> 
> aha, and what part of this PR is not "interest in the community"?

Dynamic pinctrl was asked by multiple people and was a recurrent topic, as far as I can remember. So by "community" I mean multiple parties, not just a single company.

> 
> You have two SKUs that share 99.9% of the HW but assume there are some differences in how certain devices are connected to power, so device A which is part of domain A in SKU 1 is part of domain B in SKU 2. 

Sticking to this definition, you must know how they are connected, so you can have 2 board revisions: SKU1 and SKU2.

> You can discover what SKU you are running on and adjust things accordingly at runtime.

Sounds like the same case as pinctrl, yes. For pinctrl you could use the dynamic version, true, but you could also have board revisions. As I said, dynamic pinctrl is a contained solution, the proposal here has a wider impact, which is what worries me. I think we should have a discussion on what are the limits of _dynamic_ stuff because current Zephyr foundations go into another direction.

For this case, I guess you could have a power domain driver that can be configurable at runtime based on input readings. So all devices are connected to it, but what it does is decided at runtime (I miss usecase information to understand if this is a valid option).

nashif commented:
> Dynamic pinctrl was asked by multiple people and was a recurrent topic, as far as I can remember. So by "community" I mean multiple parties, not just a single company.

Maybe we should drop x86 architecture, what the hell is this architecture doing in zephyr if it is used only by one company.

Please do not make up policies and rules as we go, we are better than this and zephyr would not made it so far following this logic.

Most of the APIs and usages came from ONE company and evolved. Something that is used and advanced by one company today moves forward and is re-used, the examples of this are countless.

nashif commented:
> Sticking to this definition, you must know how they are connected, so you can have 2 board revisions: SKU1 and SKU2.

No I can't.

ceolin commented:
> > > #42020 introduces a similar API to Linux PM QoS (simpler and part of policy manager because of Zephyr design), so it's not anything new in the end.
> > 
> > 
> > Confused how #42020 is relevant here.
> 
> It introduces a new API, with no in-tree users (yet), that's the only relation really.
> 
> > > Regarding dynamic pinctrl, it was added because a few people asked for it multiple times, so there was an interest in the community.
> > 
> > 
> > aha, and what part of this PR is not "interest in the community"?
> 
> Dynamic pinctrl was asked by multiple people and was a recurrent topic, as far as I can remember. So by "community" I mean multiple parties, not just a single company.
> 
> > You have two SKUs that share 99.9% of the HW but assume there are some differences in how certain devices are connected to power, so device A which is part of domain A in SKU 1 is part of domain B in SKU 2.
> 
> Sticking to this definition, you must know how they are connected, so you can have 2 board revisions: SKU1 and SKU2.
> 
> > You can discover what SKU you are running on and adjust things accordingly at runtime.
> 
> Sounds like the same case as pinctrl, yes. For pinctrl you could use the dynamic version, true, but you could also have board revisions. As I said, dynamic pinctrl is a contained solution, the proposal here has a wider impact, which is what worries me. I think we should have a discussion on what are the limits of _dynamic_ stuff because current Zephyr foundations go into another direction.
> 
> For this case, I guess you could have a power domain driver that can be configurable at runtime based on input readings. So all devices are connected to it, but what it does is decided at runtime (I miss usecase information to understand if this is a valid option).

But then you are not able to re-use existent power domains and you will need to re-implement it for all your domains.

gmarull commented:
> > Dynamic pinctrl was asked by multiple people and was a recurrent topic, as far as I can remember. So by "community" I mean multiple parties, not just a single company.
> 
> Maybe we should drop x86 architecture, what the hell is this architecture doing in zephyr if it is used only by one company.
> 

I'd be the first to complain about this removal because this clearly benefits the project and many companies can benefit from x86 support, I'm not sure how that compared to what we're discussing here. We're talking about technical stuff: the implications of adding dynamic stuff when Zephyr device model seems to be faced into another direction.

> Please do not make up policies and rules as we go, we are better than this and zephyr would not made it so far following this logic.
> 

I'm sorry, but I'm not making any policy here.

> Most of the APIs and usages came from ONE company and evolved. Something that is used and advanced by one company today moves forward and is re-used, the examples of this are countless.

And you'll find countless examples of features that get pushed back because the community doesn't think they're right, even though one company thinks it is.

gmarull commented:
> 
> But then you are not able to re-use existent power domains and you will need to re-implement it for all your domains.

And what is the exact problem of having, e.g., a new driver or an out of tree driver that supports a runtime configurable domain? Most projects I see have out of tree drivers, really. It could also be upstreamed if it is generic enough.

ceolin commented:
> > But then you are not able to re-use existent power domains and you will need to re-implement it for all your domains.
> 
> And what is the exact problem of having, e.g., a new driver or an out of tree driver that supports a runtime configurable domain? Most projects I see have out of tree drivers, really. It could also be upstreamed if it is generic enough.

Almost everything can be done out of tree ... Again, there is not criteria everything done in #42040 could be done in a "custom policy" that is out of tree or in a new policy and not the default and it could be upstreamed later. It seems that this will be an endless conversation, you are pointing and asking things here that yourself has not done in other situations and that is very frustrating.

ceolin commented:
> > > But then you are not able to re-use existent power domains and you will need to re-implement it for all your domains.
> > 
> > 
> > And what is the exact problem of having, e.g., a new driver or an out of tree driver that supports a runtime configurable domain? Most projects I see have out of tree drivers, really. It could also be upstreamed if it is generic enough.
> 
> Almost everything can be done out of tree ... Again, there is not criteria. Almost everything done in #42040 could be done in a "custom policy" that is out of tree or in a new policy and not the default and it could be upstreamed later. It seems that this will be an endless conversation, you are pointing and asking things here that yourself has not done in other situations and that is very frustrating.



> > > But then you are not able to re-use existent power domains and you will need to re-implement it for all your domains.
> > 
> > 
> > And what is the exact problem of having, e.g., a new driver or an out of tree driver that supports a runtime configurable domain? Most projects I see have out of tree drivers, really. It could also be upstreamed if it is generic enough.
> 
> Almost everything can be done out of tree ... Again, there is not criteria. Almost everything done in #42040 could be done in a "custom policy" that is out of tree or in a new policy and not the default and it could be upstreamed later. It seems that this will be an endless conversation, you are pointing and asking things here that yourself has not done in other situations and that is very frustrating.

And, please don't mark prs to be reviewed in meetings few hours before the meeting happens because the owner may not attending the meeting to talk about the change. Also, specific to this PR, there is a regular meeting to discuss power management that is the natural place to discuss power management first.

JordanYates commented:
I'm going to come back to a variation of my original suggestion, and propose moving this to custom behaviour of https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/power_domain/power_domain_gpio.c

This PR is focused entirely on power domains, so why not extend that driver, which is the only power domain currently in tree.
The existing build infrastructure stays the same, the handles for every device don't need to move to RAM, and you don't need to bloat the handles array for every other device. If we end up having multiple power domains in tree, the functionality could be added to a common `power_domain_api`.

ceolin commented:
> I'm going to come back to a variation of my original suggestion, and propose moving this to custom behaviour of https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/power_domain/power_domain_gpio.c
> 
> This PR is focused entirely on power domains, so why not extend that driver, which is the only power domain currently in tree.
> The existing build infrastructure stays the same, the handles for every device don't need to move to RAM, and you don't need to bloat the handles array for every other device. If we end up having multiple power domains in tree, the functionality could be added to a common `power_domain_api`.

Why I would add this API for a power domain that is not used in our use case ? Sorry but this is not generic for power domains, the use case is to be able to map a device to any sort of power domain and we would end up needing to re-implement for all different power domains. Once again, the feature is completely optional as has no impact for those who not want to use it

JordanYates commented:
> Why I would add this API for a power domain that is not used in our use case ? Sorry but this would be specific for this power domain, the use case is to be able to map a device to any sort of power domain and we would end up needing to re-implement for all different power domains.

There are options beyond just copy pasting an implementation into N different drivers. 

> Once again, the feature is completely optional as has no impact for those who not want to use it

Personally what I'm not a fan of is that enabling a feature for a single type of device changes device generation and memory layout globally. I feel like there are more targeted solutions. I will add come comments of what I think would improve this if the global path is still chosen.

JordanYates commented on include/linker/common-ram.ld:
```diff
@@ -45,6 +45,19 @@
+#if defined(CONFIG_PM_DEVICE_POWER_DOMAIN_DYNAMIC) && (CONFIG_PM_DEVICE_POWER_DOMAIN_DYNAMIC != 0)
+	SECTION_DATA_PROLOGUE(device_handles,,)
+	{
+		__device_handles_start = .;
```
This common section for linking the device handles in ROM or RAM could be pulled out to a common file (device-handles.ld) that is included in both locations so we can't update one and miss the other

ceolin replied:
let me look how it will looks like and come up with a new version.

JordanYates commented on subsys/pm/Kconfig:
```diff
@@ -82,6 +82,14 @@ config PM_DEVICE_POWER_DOMAIN
 	  devices that depend on a domain will be notified when this
 	  domain is suspended or resumed.
 
+config PM_DEVICE_POWER_DOMAIN_DYNAMIC
```
Make the build system updates less specific to this particular feature. Add a (hidden?) option (`config DEVICE_DYNAMIC_HANDLES` or similar) which this symbol can select and can drive the changes in `device.h` and the linker scripts.

ceolin replied:
That was something that I initially had thought and decided to take the current path to avoid and additional symbol, but that is something I definitely can change.

JordanYates reviewed (dismissed):
The proposed changes make this less objectionable to me, removing duplication in the linker scripts and cleaning up the conditionals in `device.h`

gmarull commented:
So then, why isn’t it possible to implement a custom PM domain driver instead of a more invasive solution like the one presented here? Or use board revisions? I think these question have not been clearly answered yet.

ceolin commented:
> So then, why isn’t it possible to implement a custom PM domain driver instead of a more invasive solution like the one presented here? Or use board revisions? I think these question have not been clearly answered yet.

Can you deploy the same binary (without re-compile) in two different targets using board revisions ? If the answer is no you have your answer. Because that is a behavior that is useful for any power domain. Let me be clear, one more time, it can be done in custom driver, but we would end up having to do it for our custom pm_driver_x, the custom_pm_driver_y, so on a so forth. It is not like all the pm drivers would do the same thing .... This is not invasive,  .... it does not change anything if it is disabled and it it is not a major change.

These questions were answered many times in this discussion, there is a real use case for this feature. Now what I getting here is, I don't like the feature and I don't care if there are users asking for it, simply because I don't care ...

ceolin commented:
> > Why I would add this API for a power domain that is not used in our use case ? Sorry but this would be specific for this power domain, the use case is to be able to map a device to any sort of power domain and we would end up needing to re-implement for all different power domains.
> 
> There are options beyond just copy pasting an implementation into N different drivers.
> 
> > Once again, the feature is completely optional as has no impact for those who not want to use it
> 
> Personally what I'm not a fan of is that enabling a feature for a single type of device changes device generation and memory layout globally. I feel like there are more targeted solutions. I will add come comments of what I think would improve this if the global path is still chosen.

I can resize only devices that are domains, and so not increase memory usage for other devices ... This of course will require more complicated changed. Doing it in an additional commit so we can review it.

Commit:
pm: Add an underline in device pm slot

Add an underline character before "slot" to make the sufix easier to
read.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: device: Use pm flags to mark a power domain

Add a new flag that is used to tell whether or not a device is a power
domain. In order to a device be identified as a power domain the node
has to be compatible with "power-domain".

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: Only resize power domains

Instead of resizing all devices handles, we just resize devices that are
power domains. This means that a power domain has to be declared as
compatbile with "power-domain" in device tree node.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

carlocaione commented:
I will join the club of those who do not like this for two reasons:

- Everything of this commit is hidden deep inside the Zephyr core code. There is no DT exposure at all and all the magic is done in some automatic way behind the curtains in a way that is kinda special and custom. And yes, this is inconsistent with very much all the other zephyr components.

- To me it looks like this PR is taking shortcuts. Why this is not implemented as a new driver or subsystem or framework with a proper set of changes well compartmentalized?

ceolin commented:
> I will join the club of those who do not like this for two reasons (at least):
> 
> * Everything of this commit is hidden deep inside the Zephyr core code. There is no DT exposure at all and all the magic is done in some automatic way behind the curtains in a way that is kinda special and custom. And yes, this is inconsistent with very much all the other zephyr components.

The logic behind the handles is probably is why this is deep inside the code. We have that magic happen in build time and in order to reserve some slots we have to change there. That is not allocating anything dynamically, it is about to configure / reconfigure something at runtime.

> * To me it looks like this PR is taking shortcuts. Why this is not implemented as a new driver or subsystem or framework with a proper set of changes well compartmentalized?

The reason is because this is a behavior that applies for any power domain and it is not particular to one driver. Another use case that would take advantage of this feature is when you have a dma controller that is reconfigured to do a transfer in a different bus. In this case it makes sense to bind the controller to the power domain that the bus is in.

nashif commented:
> * Everything of this commit is hidden deep inside the Zephyr core code.

isn't this how Zephyr works in general? All the userspace and syscall code generation, DT device dependencies and handles, build system automagic all over the place, so when you try to do something simple, you end up touching all those areas under the hood.

gmarull commented:
So I wonder why something like this can't be done:

```c
static int mydomain_init(const struct device *dev)
{
	struct mydomain_data *data = dev->data;
	const struct mydomain_config *config = dev->config;

	/* determine at runtime if SKU1 or 2 */
	if (is_sku1()) {
		data->gpio = config->sku1_gpio;
	} else {
		data->gpio = config->sku2_gpio;
	}

	/* configure data->gpio */
	...

	return 0;
}

#define MYDOMAIN_DEFINE(n)						       \
	static const struct mydomain_config mydomain_config##n = {	       \
		.sku1_gpio = GPIO_DT_SPEC_INST_GET(n, sku1_gpios),	       \
		.sku2_gpio = GPIO_DT_SPEC_INST_GET(n, sku2_gpios),	       \
	};								       \
									       \
	static struct mydomain_data mydomain_data##n;			       \
									       \
	DEVICE_DT_INST_DEFINE(n, &mydomain_init, NULL, &mydomain_data##n,      \
			      &mydomain_config##n, POST_KERNEL,		       \
			      0, NULL);

DT_INST_FOREACH_STATUS_OKAY(MYDOMAIN_DEFINE)
```

ceolin commented:
> So I wonder why something like this can't be done:
> 
> ```c
> static int mydomain_init(const struct device *dev)
> {
> 	struct mydomain_data *data = dev->data;
> 	const struct mydomain_config *config = dev->config;
> 
> 	/* determine at runtime if SKU1 or 2 */
> 	if (is_sku1()) {
> 		data->gpio = config->sku1_gpio;
> 	} else {
> 		data->gpio = config->sku2_gpio;
> 	}
> 
> 	/* configure data->gpio */
> 	...
> 
> 	return 0;
> }
> 
> #define MYDOMAIN_DEFINE(n)						       \
> 	static const struct mydomain_config mydomain_config##n = {	       \
> 		.sku1_gpio = GPIO_DT_SPEC_INST_GET(n, sku1_gpios),	       \
> 		.sku2_gpio = GPIO_DT_SPEC_INST_GET(n, sku2_gpios),	       \
> 	};								       \
> 									       \
> 	static struct mydomain_data mydomain_data##n;			       \
> 									       \
> 	DEVICE_DT_INST_DEFINE(n, &mydomain_init, NULL, &mydomain_data##n,      \
> 			      &mydomain_config##n, POST_KERNEL,		       \
> 			      0, NULL);
> 
> DT_INST_FOREACH_STATUS_OKAY(MYDOMAIN_DEFINE)
> ```

@gmarull The reason was said before, because this has to be done for all power domain (or another additional kind of infrastructure has to be provided to avoid copy&paste. Also, this does not solve problem of when you have a dma controller that is reconfigured to do a transfer in a different bus and you want to bind the controller to the power domain that the bus is in.

ceolin commented:
> > @gmarull The reason was said before, because this has to be done for all power domains (or another additional kind of infrastructure has to be provided to avoid copy&paste). Also, this does not solve problem of when you have a dma controller that is reconfigured to do a transfer in a different bus and you want to bind the controller to the power domain that the bus is in.
> 
> With this description, it's clear to me that there's missing usecase information here. We went from SKU1/2 to a DMA controller that is reconfigured to do X and Y. Note that dynamic pinctrl would not be able to handle such scenarios (it's limited to boot time pre-device init). So, it's definitely not the same case as some kept insisting.

@gmarull that is not the initial use case, we not went from one to another, but to a second use case ... you are not being reasonable, it was never said that the sku1/2 use case is not longer needed or exists ... I've added a second use case that was brought my attention by another co-worker.

teburd reviewed (dismissed):
Looks good to me and seems entirely opt in.

jfischer-no commented on include/pm/device.h:
```diff
@@ -483,6 +483,36 @@ bool pm_device_state_is_locked(const struct device *dev);
+ * @retval -ENOSYS If the application was built without power domain support.
+ * @retval -ENOSPC If there is no space available in the power domain to add the device.
+ */
+int pm_device_power_domain_add(const struct device *dev,
```
`int pm_dpd_add(const struct device *dev, const struct device *domain);`,
IMO the abbreviation `_dpd_` for `_device_power_domain_` is obvious.

ceolin replied:
hm it looks to mnemonic for me and does not keep the namespace, maybe `pm_device_pd_add` ?

galak commented on CMakeLists.txt:
```diff
@@ -836,6 +836,12 @@ zephyr_get_include_directories_for_lang(C
   STRIP_PREFIX # Don't use a -I prefix
   )
 
+if(CONFIG_PM_DEVICE_POWER_DOMAIN_DYNAMIC)
```
We should introduce a hidden Kconfig option for the CONFIG_DYNAMIC_DEVICES that gets selected by CONFIG_PM_DEVICE_POWER_DOMAIN_DYNAMIC.  So that the dynamic device functionality could be used by other use cases.

ceolin replied:
That makes sense

ceolin replied:
I'll just change a little bit it, `CONFIG_PM_DEVICE_POWER_DOMAIN_DYNAMIC` contains the number of slots that should be pre-allocated, since we don't have the need to resize all devices but only power domains, I'll keep the power domain name. What I'll change though is the hidden option that tells if the build supports dynamic devices. In the end this is the symbol that is used inside `device.h` and linker scripts. If I remember correctly the concern was about having PM specifics in those files.

The other possibility is to make it generic and resize all devices.

nashif replied:
the idea was to guard with CONFIG_DYNAMIC_DEVICES (which would be selected by CONFIG_PM_DEVICE_POWER_DOMAIN_DYNAMIC > 0 ) and then use CONFIG_PM_DEVICE_POWER_DOMAIN_DYNAMIC below now for power domain, but someone else might add more devices, so they feature is exposed as a generic one. Really cosmetic and can be done later when there are more users, but I guess this would make it cleaner.

ceolin replied:
that is basically what I've done AFAIU, just the name that is different.

gmarull commented on include/device.h:
```diff
@@ -877,9 +882,11 @@ __deprecated static inline int device_usable_check(const struct device *dev)
 		Z_DEVICE_HANDLE_NAME(node_id, dev_name)[];		\
-	const device_handle_t						\
+	COND_CODE_0(CONFIG_HAS_DYNAMIC_DEVICES,				\
+		    (const device_handle_t), (device_handle_t))		\
```
note: the only thing being modified here is the `const` attribute for device handles, so I think a better option would be something like:

```
#ifdef CONFIG_DEVICE_HANDLES_ROMABLE
#define Z_DEVICE_HANDLE_CONST const
#else
#define Z_DEVICE_HANDLE_CONST
#endif
```

and make `CONFIG_DEVICE_HANDLES_ROMABLE` a hidden option selected by PM domains (name may be changed, it's the first that came to my mind). Then, just use `Z_DEVICE_HANDLE_CONST` instead of `const` everywhere.

p.d. I did something similar for pinctrl https://github.com/zephyrproject-rtos/zephyr/blob/main/include/drivers/pinctrl.h#L194-L204

ceolin replied:
I liked the idea, just the new Kconfig symbol that I don't think it is necessary. We already have the HAS_DYNAMIC_DEVICE or whatever other name we come up. Basically,
```
#ifdef CONFIG_HAS_DYNAMIC_DEVICES
#define Z_DEVICE_HANDLE_CONST const
#else
#define Z_DEVICE_HANDLE_CONST
#endif
```

gmarull replied:
I'm still confused on what a "dynamic device" is

ceolin replied:
forgot the name, it can be changed. I just mentioned that I don't want to introduce a redundant symbol only for the constant attribute. We should be able to re-use the same symbol everywhere, `HANDLES_ROMABLE` does not say anything about why we have this condition as well. These hidden symbols are pretty hard to maintain and also make the code more difficult to maintain. One hidden symbol for this feature is enough.

gmarull commented on include/linker/common-ram.ld:
```diff
@@ -45,6 +45,19 @@
 	} GROUP_DATA_LINK_IN(RAMABLE_REGION, ROMABLE_REGION)
 #endif
 
+#if defined(CONFIG_HAS_DYNAMIC_DEVICES)
+	SECTION_DATA_PROLOGUE(device_handles,,)
+	{
+	__device_handles_start = .;
+#ifdef LINKER_DEVICE_HANDLES_PASS1
+	KEEP(*(SORT(.__device_handles_pass1*)));
+#else
+	KEEP(*(SORT(.__device_handles_pass2*)));
+#endif /* LINKER_DEVICE_HANDLES_PASS1 */
+	__device_handles_end = .;
+	} GROUP_DATA_LINK_IN(RAMABLE_REGION, ROMABLE_REGION)
+#endif 	/* CONFIG_HAS_DYNAMIC_DEVICES */
```
note: would it be possible to avoid duplication?

ceolin replied:
I tried to split it in a different file, the problem is that the file is included in both linker script fragments, `commo-ram.ld` and `common-rom.ld`. Without know which fragment the file is being included the symbol `CONFIG_HAS_DYNAMIC_DEVICES` is not enough. e.g:
```
#if defined(CONFIG_HAS_DYNAMIC_DEVICES)
	SECTION_DATA_PROLOGUE(device_handles,,)
	{
	__device_handles_start = .;
#ifdef LINKER_DEVICE_HANDLES_PASS1
	KEEP(*(SORT(.__device_handles_pass1*)));
#else
	KEEP(*(SORT(.__device_handles_pass2*)));
#endif /* LINKER_DEVICE_HANDLES_PASS1 */
	__device_handles_end = .;
	} GROUP_DATA_LINK_IN(RAMABLE_REGION, ROMABLE_REGION)
#else /* No dynamic devices */
	SECTION_DATA_PROLOGUE(device_handles,,)
	{
	__device_handles_start = .;
#ifdef LINKER_DEVICE_HANDLES_PASS1
	KEEP(*(SORT(.__device_handles_pass1*)));
#else
	KEEP(*(SORT(.__device_handles_pass2*)));
#endif /* LINKER_DEVICE_HANDLES_PASS1 */
	__device_handles_end = .;
	} GROUP_ROM_LINK_IN(RAMABLE_REGION, ROMABLE_REGION)
#endif
```

This would be end up defining __device_handles_start in both ram and rom independently of `CONFIG_HAS_DYNAMIC_DEVICES` value.  The snippet above can be simplified, but does not solve the problem.

JordanYates replied:
My original proposal was to strip out the internals of the 
```
SECTION_DATA_PROLOGUE()
{
   /* Strip everything in here */
} > GROUP_LINK_IN()
```
into the common file, leaving the `#if defined(CONFIG_HAS_DYNAMIC_DEVICES)` where it is.
Just to keep the rather messy device handle pass logic in only one location.

ceolin replied:
ok, that is possible.

gmarull commented on kernel/Kconfig:
```diff
@@ -928,4 +928,16 @@ config THREAD_LOCAL_STORAGE
 
+menu "Device Options"
+
+config HAS_DYNAMIC_DEVICES
```
as I mentioned, I think the "dynamic devices" term is confusing. Maybe it's because it's not explained anywhere what a "dynamic device" is.

ceolin replied:
that is fair, what a about `DYNAMIC_DEVICE_DEPENDENCIES`  ?

ceolin replied:
changed to `HAS_DYNAMIC_DEVICE_HANDLES`

gmarull commented on subsys/pm/device.c:
```diff
@@ -101,6 +101,82 @@ int pm_device_action_run(const struct device *dev,
+	}
+
+	return add ? -ENOSPC : -ENOENT;
+#else
+	ARG_UNUSED(dev);
+	ARG_UNUSED(domain);
+	ARG_UNUSED(add);
+
+	return -ENOSYS;
+#endif
```
why is this branch needed? `pm_device_power_domain_remove/add` already return `-ENOSYS` from a header inline right?

ceolin replied:
the function declaration is under `CONFIG_PM_DEVICE` guard. You can have CONFIG_PM_DEVICE` enabled and not have this feature. That is why. I can, make it fully inline in the header, it just make the header more complex. There is another function doing the same thing, `pm_device_on_power_domain()`

gmarull commented on tests/subsys/pm/power_domain/prj.conf:
```diff
@@ -2,5 +2,6 @@ CONFIG_ZTEST=y
 CONFIG_PM=y
 CONFIG_PM_DEVICE=y
 CONFIG_POWER_DOMAIN=y
+CONFIG_PM_DEVICE_POWER_DOMAIN_DYNAMIC=1
```
```suggestion
CONFIG_PM_DEVICE_POWER_DOMAIN_DYNAMIC=y
```

ceolin replied:
No, this breaks the logic, that symbols is the one that is used to check how many slots should be pre-allocated.

gmarull replied:
so then I'd make the option name contain `_NUM` somewhere

gmarull reviewed (commented):
added some more comments, will re-review again

ceolin commented:
> note: the only thing being modified here is the `const` attribute for device handles, so I think a better option would be something like:

```
#ifdef CONFIG_DEVICE_HANDLES_ROMABLE
#define Z_DEVICE_HANDLE_CONST const
#else
#define Z_DEVICE_HANDLE_CONST
#endif
```

and make `CONFIG_DEVICE_HANDLES_ROMABLE` a hidden option selected by PM domains (name may be changed, it's the first that came to my mind). Then, just use `Z_DEVICE_HANDLE_CONST` instead of `const` everywhere.

p.d. I did something similar for pinctrl https://github.com/zephyrproject-rtos/zephyr/blob/main/include/drivers/pinctrl.h#L194-L204 

I liked the idea, just the new Kconfig symbol that I don't think it is necessary. We already have the  `HAS_DYNAMIC_DEVICE` or whatever other name we come up. Basically,

```
#ifdef CONFIG_HAS_DYNAMIC_DEVICES
#define Z_DEVICE_HANDLE_CONST const
#else
#define Z_DEVICE_HANDLE_CONST
#endif
```

ceolin commented:
> ```suggestion
CONFIG_PM_DEVICE_POWER_DOMAIN_DYNAMIC=y
``` 
No, this breaks the logic, that symbols is the one that is used to check how many slots should be pre-allocated.

Commit:
linker: Re-use device handle logic

Add a new file with the device handles logic that is included in both
ram and ROM

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

gmarull commented on subsys/pm/Kconfig:
```diff
@@ -82,6 +82,14 @@ config PM_DEVICE_POWER_DOMAIN
 	  domain is suspended or resumed.
 
+config PM_DEVICE_POWER_DOMAIN_DYNAMIC_NUM
+	int "Enable dynamic binding device to a power domain"
```
```suggestion
	int "Number of devices dynamically bindable to a power domain"
```

gmarull commented on CMakeLists.txt:
```diff
@@ -845,6 +851,7 @@ if(CONFIG_HAS_DTS)
     ${PYTHON_EXECUTABLE}
     ${ZEPHYR_BASE}/scripts/gen_handles.py
     --output-source dev_handles.c
+    --dynamic-devices ${number_of_dynamic_devices}
```
nit:
```suggestion
    --num-dynamic-devices ${number_of_dynamic_devices}
```

gmarull commented on scripts/gen_handles.py:
```diff
@@ -154,6 +158,55 @@ def obj_handles(self):
+
+class PMDevice:
+    """
+    Represents information about a pm_odevice object and its references to other objects.
```
```suggestion
    Represents information about a pm_device object and its references to other objects.
```

gmarull commented on kernel/Kconfig:
```diff
@@ -928,4 +928,16 @@ config THREAD_LOCAL_STORAGE
+	depends on (PM_DEVICE_POWER_DOMAIN_DYNAMIC_NUM > 0)
+	default y
+	help
+	  Hidden option to enable support for dynamically add devices
+	  to supported handles.
```
```suggestion
	  Hidden option that makes device handles non-const. It can be useful when device handles have to be manipulated at runtime.
```

ceolin replied:
`Hidden option that makes possible to manipulate device handles at  runtime.`

I think the non-const is not necessary, that is not the only thing that this option does, so instead of partially describe the implementation I rather focus in the behavior / feature.

gmarull commented on kernel/Kconfig:
```diff
@@ -928,4 +928,16 @@ config THREAD_LOCAL_STORAGE
+
+config HAS_DYNAMIC_DEVICE_HANDLES
+	bool
+	depends on (PM_DEVICE_POWER_DOMAIN_DYNAMIC_NUM > 0)
+	default y
```
hidden options should be selected from somewhere else, we're adding here PM subsys dependencies into `kernel/Kconfig`.

ceolin replied:
ok, that is fair. It will necessary an additional symbol to select this one I believe though

gmarull commented on kernel/Kconfig:
```diff
@@ -928,4 +928,15 @@ config THREAD_LOCAL_STORAGE
+
+config HAS_DYNAMIC_DEVICE_HANDLES
+	bool
+	default n
```
nit: redundant

gmarull commented on subsys/pm/Kconfig:
```diff
@@ -82,6 +82,24 @@ config PM_DEVICE_POWER_DOMAIN
 	  devices that depend on a domain will be notified when this
 	  domain is suspended or resumed.
 
+config PM_DEVICE_POWER_DOMAIN_DYNAMIC_NUM
+	int "Number of devices that can dynamically be bind to a power domain"
+	depends on PM_DEVICE_POWER_DOMAIN
+	default 0
+	help
+	  This option allows dynamically bind up to the given value
+	  number of devices to a Power Domain.
+
+# Since it is not possible to select HAS_DYNAMIC_DEVICE_HANDLES
+# directly from PM_DEVICE_POWER_DOMAIN_DYNAMIC_NUM, we have
+# this hidden symbol just to select HAS_DYNAMIC_DEVICE_HANDLES if
+# the number of power domain dynamic devices is greater than 0
+config HAS_PM_DEVICE_POWER_DOMAIN_DYNAMIC
+	bool
+	depends on PM_DEVICE_POWER_DOMAIN_DYNAMIC_NUM > 0
+	default y
+	select HAS_DYNAMIC_DEVICE_HANDLES
```
I think it would be better to do something like this:

```
config PM_DEVICE_POWER_DOMAIN_DYNAMIC
	bool "Allow to dynamically bind devices to a power domain"
	depends on PM_DEVICE_POWER_DOMAIN
	select HAS_DYNAMIC_DEVICE_HANDLES
	help
	  ...

config PM_DEVICE_POWER_DOMAIN_DYNAMIC_NUM
	int "Number of devices that can dynamically be bind to a power domain"
	depends on PM_DEVICE_POWER_DOMAIN_DYNAMIC
	default 0 /* maybe can be set to 1 at least with suggested approach approach, as PM_DEVICE_POWER_DOMAIN_DYNAMIC_NUM is no longer used for boolean checks  */
	help
	  This option allows dynamically bind up to the given value
	  number of devices to a Power Domain.
```

ceolin replied:
I think that is better indeed, more verbose since requires enable the feature and then select the number of slots, but it is clearer. Will change it.

gmarull commented on kernel/Kconfig:
```diff
@@ -928,4 +928,15 @@ config THREAD_LOCAL_STORAGE
 
+menu "Device Options"
+
+config HAS_DYNAMIC_DEVICE_HANDLES
```
nit: `HAS` may be changed to something else for clarity (we don't have anything, it's an option to select a requirement), maybe `USE` or `NEEDS` (not sure what if there's a pattern).

ceolin replied:
There are some places using `NEEDS` but the majority of similar cases are using `HAS`.

gmarull commented on CMakeLists.txt:
```diff
@@ -840,6 +840,12 @@ zephyr_get_include_directories_for_lang(C
   STRIP_PREFIX # Don't use a -I prefix
   )
 
+if(CONFIG_PM_DEVICE_POWER_DOMAIN_DYNAMIC_NUM)
```
```suggestion
if(CONFIG_PM_DEVICE_POWER_DOMAIN_DYNAMIC)
```

ceolin replied:
done

gmarull reviewed (dismissed):
one minor nitpick, otherwise lgtm (note that this approach may be revisited if one day Zephyr has better support for "dynamic devices")

ceolin commented:
@nashif @gmarull can you re-approve it please ?
